elements/dash-tree/dash-tree.js

```diff
@@ -0,0 +1,448 @@
+ * should be a `DashTreeGroup` element, which will hold all of its
+ * `DashTreeItem` children.
+ *
+ * Parent nodes can be either collapsed or expanded to reveal their children.
+ * The state of the parent node is conveyed through the use of a `aria-expanded`
+ * attribute.
+ *
+ * Depending on the implementation, trees can support either single or
+ * multi selection. The `DashTree` element supports single selection, so
+ * there can only be one selected element at a time. The currently selected
+ * element is indicated by the `aria-selected` attribute.
+ *
+ * Unlike the `DashRadioGroup`, which uses roving tabindex to indicate which
+ * child is currently active, the `DashTree` uses `aria-activedescendant` and
+ * the `id` of the currently active child. The effect is similar to using roving
+ * tabindex, and is presented in this case to show an alternative approach to
+ * indicating active children.
+ */
+
+(function() {
+  /**
+   * Define keycodes to help with handling keyboard events.
+   */
+  const KEYCODE = {
+    UP: 38,
+    DOWN: 40,
+    LEFT: 37,
+    RIGHT: 39,
+    HOME: 36,
+    END: 35,
+    SPACE: 32,
+    ENTER: 13,
+  };
+
+  /**
+   * A helper to quickly identify `DashTreeItem` nodes
+   */
+  function isTreeItem(node) {
+    return node.nodeName.toLowerCase() === 'dash-treeitem';
+  }
+
+  /**
+   * A helper to quickly identify ARIA states.
+   * Note that ARIA attributes require a literal string value of "true"/"false".
+   */
+  function isExpanded(node) {
+    return node.getAttribute('aria-expanded') === 'true';
+  }
+
+  // `dashTreeItemCounter` counts the number of `<dash-treeitem>` instances
+  // created. The number is used to generated new, unique IDs.
+  let dashTreeItemCounter = 0;
+
+  class DashTreeItem extends HTMLElement {
+    constructor() {
+      super();
+    }
+
+    connectedCallback() {
+      this.setAttribute('role', 'treeitem');
+
+      // If the element doesn't already have an `id`, generate one for it.
+      // This will make it easier to set the element as active using
+      // `aria-activedescendant`.
+      if (!this.id)
+        this.id = `dash-treeitem-generated-${dashTreeItemCounter++}`;
+
+      // If the element contains a `<dash-treegroup>` then it's a parent node
+      if (this.querySelector('dash-treegroup')) {
+        // If the element is not explicitly already expanded by the user, then
+        // set it to closed.
+        if (!isExpanded(this))
+          this.setAttribute('aria-expanded', false);
+
+        // Set the `aria-label` to the first bit of text. Otherwise the label
+        // for this element will be computed based on its text content plus
+        // the text content of all of its children. Making it so verbose as to
+        // be unusable. This first child should ideally be a `<span>` element,
+        // but it may also just be a line of text.
+        if (!this.hasAttribute('aria-label')) {
+          let label = this.querySelector('span');
```

Comment:
yeah not sure to be honest. `<label>` only works with specific input elements, select and textarea. But it may be nicer to use it since it semantically conveys what we're going for. And hopefully a future version of Custom Elements will be able to work with label elements. I'll add it in and see how Alice feels :D